File: ZMQ/lec-05-prg-06-pub-sub-and-pull-push-client.py
# Rerefence: https://stackoverflow.com/questions/49289072/zmq-dealer-router-communication

import random
import time

import zmq

def main():

    # Prepare our context and subscriber
    ctx = zmq.Context()
    subscriber = ctx.socket(zmq.SUB)
    subscriber.setsockopt(zmq.SUBSCRIBE, b'')
    subscriber.connect("tcp://localhost:5557")
    publisher = ctx.socket(zmq.PUSH)
    publisher.connect("tcp://localhost:5558")

    random.seed(time.time())
    while True:
        if subscriber.poll(100) & zmq.POLLIN:
            message = subscriber.recv()
            print("I: received message ", message)
        else:
            rand = random.randint(1, 100)
            if rand < 10:
                publisher.send(b"%d" % rand)
                print("I: sending message ", rand)

if __name__ == '__main__':
    main()

File: ZMQ/lec-05-prg-06-pub-sub-and-pull-push-client.js
const zmq = require("zeromq");

async function run() {
  const subscriber = new zmq.Subscriber();
  subscriber.connect("tcp://localhost:5557");
  subscriber.subscribe();

  const publisher = new zmq.Push();
  publisher.connect("tcp://localhost:5558");

  setInterval(() => {
    const rand = Math.floor(Math.random() * 100);
    if (rand < 10) {
      publisher.send(rand);
      console.log("I: sending message ", rand);
    }
  }, 250);

  while (true) {
    const message = await subscriber.receive();
    console.log("I: received message ", message.toString());
  }
}

run();
